tracer/src/Datadog.Trace/Debugger/Caching/LFUEvictionPolicy.cs

```diff
@@ -0,0 +1,142 @@
+                    var now = DateTime.UtcNow;
+                    _frequencyMap[key] = new FrequencyItem(1, now);
+                    AddToFrequencySet(key, 1, now);
+                }
+            }
+            finally
+            {
+                _lock.ExitWriteLock();
+            }
+        }
+
+        public void Remove(TKey key)
+        {
+            _lock.EnterWriteLock();
+            try
+            {
+                if (_frequencyMap.TryGetValue(key, out var item))
+                {
+                    RemoveFromFrequencySet(key, item.Frequency, item.LastAccessed);
+                    _frequencyMap.Remove(key);
+                }
+            }
+            finally
+            {
+                _lock.ExitWriteLock();
+            }
+        }
+
+        public void Access(TKey key)
+        {
+            _lock.EnterWriteLock();
+            try
+            {
+                if (_frequencyMap.TryGetValue(key, out var item))
+                {
+                    RemoveFromFrequencySet(key, item.Frequency, item.LastAccessed);
+                    item.Frequency++;
+                    item.LastAccessed = DateTime.UtcNow;
+                    AddToFrequencySet(key, item.Frequency, item.LastAccessed);
+                }
+            }
+            finally
+            {
+                _lock.ExitWriteLock();
+            }
+        }
+
+        public TKey Evict()
+        {
+            _lock.EnterWriteLock();
+            try
+            {
+                if (_frequencySortedSet.Count == 0)
+                {
+                    throw new InvalidOperationException("Cache is empty");
+                }
+
+                var leastFrequent = _frequencySortedSet.First();
+                var key = leastFrequent.Value.First();
+                Remove(key);
+                return key;
+            }
+            finally
+            {
+                _lock.ExitWriteLock();
+            }
+        }
+
+        private void AddToFrequencySet(TKey key, int frequency, DateTime lastAccessed)
+        {
+            var tuple = new FrequencyKey(frequency, lastAccessed);
+            if (!_frequencySortedSet.TryGetValue(tuple, out var set))
+            {
+                set = new HashSet<TKey>();
+                _frequencySortedSet[tuple] = set;
+            }
+
+            set.Add(key);
+        }
+
+        private void RemoveFromFrequencySet(TKey key, int frequency, DateTime lastAccessed)
+        {
+            var tuple = new FrequencyKey(frequency, lastAccessed);
+            if (_frequencySortedSet.TryGetValue(tuple, out var set))
+            {
+                set.Remove(key);
+                if (set.Count == 0)
+                {
+                    _frequencySortedSet.Remove(tuple);
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            _lock.Dispose();
+        }
+
+        private record struct FrequencyKey(int Frequency, DateTime LastAccessed);
```

Comment:
For my understanding, what's the reason we are using a struct for the FrequencyKey and a class for the FrequencyItem?

Reply:
While `FrequencyKey` is immutable object for lookup/sorting in adictionary, `FrequencyItem` is mutable object and doing it as a struct would cause defensice copies when modifying his properties